cptv-download.py

```diff
@@ -0,0 +1,238 @@
+        file_count = 0
+        for row in rows:
+            file_count += 1
+            pool.put(row)
+        pool.stop()
+        print("Finished {0} files.".format(file_count))
+
+    def update_file_locations(self):
+        """ Scans output folder building a list of all files. """
+        self.file_list = {}
+
+        for root, subdirs, files in os.walk(self.out_folder):
+            for file in files:
+                if file not in self.file_list:
+                    self.file_list[file] = []
+                self.file_list[file].append(root)
+
+    def _downloader(self, q, api, out_base):
+        """ Worker to handle downloading of files. """
+        while True:
+            r = q.get()
+
+            if r is None:
+                break
+
+            try:
+                tag_dir = get_tag_directory(r['Tags'])
+
+                out_dir = os.path.join(out_base, tag_dir)
+                dt = parsedate(r['recordingDateTime'])
+                file_base = dt.strftime("%Y%m%d-%H%M%S") + "-" + r['Device']['devicename']
+
+                path_base = os.path.join(out_dir, file_base)
+
+                if self.auto_delete:
+                    if file_base+'.cptv' in self.file_list:
+                        for existing_path in self.file_list[file_base+'.cptv']:
+                            if existing_path != out_dir and os.path.basename(existing_path) not in ['test','hard']:
+                                print("Found {} in {} but should be in {}".format(file_base, existing_path, out_dir))
+                                remove_file(os.path.join(existing_path, file_base + '.cptv'))
+                                remove_file(os.path.join(existing_path, file_base + '.dat'))
+                                remove_file(os.path.join(existing_path, file_base + '.mp4'))
+
+                if tag_dir in self.ignore_tags:
+                    continue
+
+                if self.only_tags and tag_dir not in self.only_tags:
+                    continue
+
+                if not os.path.exists(out_dir):
+                    os.makedirs(out_dir)
+
+                if iter_to_file(path_base + '.cptv', api.download_cptv(r['id'])):
+                    print(format_row(r) + '.cptv' + " [{}]".format(tag_dir))
+
+                if self.include_mp4:
+                    if iter_to_file(path_base + '.mp4', api.download_mp4(r['id'])):
+                        print(format_row(r) + '.mp4' + " [{}]".format(tag_dir))
+
+                if self.include_metadata:
+                    if not os.path.exists(path_base+'.txt'):
+                        json.dump(r, open(path_base+'.txt', 'w'), indent=4)
+
+            finally:
+                q.task_done()
+
+
+def remove_file(file):
+    """ Delete a file (if it exists). """
+    if os.path.exists(file):
+        try:
+            os.remove(file)
+        except Exception as e:
+            print("Warning, could not remove file {}. Error: {}".format(file, e))
```

Comment:
Instead of "looking before you leap" it's better to just try to do the thing (removing a file in this case) and then handle any exception that results. Python 3 will raise a FileNotFoundError if the file isn't there (which should just be handled with a `pass`). An OSError should result in the print and anything else should be left unhandled as this is truly exceptional.

Reply:
good idea!